File: src-tauri/src/events.rs
//! Outbound event emit helpers (Rust -> webview, SPEC s11.7).
//!
//! Thin wrappers over the Tauri v2 [`Emitter`] trait (`app.emit(event,
//! payload)`; the v1 `emit_all` was removed - SPEC s11.7). Each helper pins
//! the canonical SPEC s11.7 event name so call sites cannot typo the channel.
//! Payload DTOs (`GlobalSyncStatus`, `ActivityEntry`) land with the IPC layer
//! in M6/M7; until then the broadcast helpers are generic over any
//! `serde::Serialize` payload so the orchestrator-bridge can emit without the
//! DTO crate existing yet.

use serde::Serialize;
use tauri::{AppHandle, Emitter};

/// `sync:status_changed` - global sync status changed (payload:
/// `GlobalSyncStatus`, SPEC s11.7).
pub const EVENT_SYNC_STATUS_CHANGED: &str = "sync:status_changed";
/// `sync:source_progress` - per-source progress (payload:
/// `{ source_id, progress }`, SPEC s11.7).
///
/// Reserved for M6: the per-source progress DTO + the bridge that emits it land
/// with the M6 IPC layer (the M5 event bridge only forwards `StateChanged`).
#[allow(dead_code)]
pub const EVENT_SYNC_SOURCE_PROGRESS: &str = "sync:source_progress";
/// `activity:new` - a new activity-log entry (payload: `ActivityEntry`,
/// SPEC s11.7).
///
/// M7 (activity dashboard): the orchestrator broadcasts
/// `OrchestratorEvent::ActivityWritten` on every durable activity write and the
/// event bridge re-emits the carried `ActivityEntry` on this channel for the
/// dashboard's live tail.
pub const EVENT_ACTIVITY_NEW: &str = "activity:new";
/// `activity:lagged` - the live-tail broadcast lagged and dropped one or more
/// `activity:new` events (payload: none).
///
/// M7-P1-1: the per-account `OrchestratorEvent` broadcast is bounded (capacity
/// 256). Under a burst (e.g. a per-file error storm) the event bridge's receiver
/// can lag and the dropped events would be PERMANENTLY missing from the live
/// tail, violating DESIGN s8.3's last-1000 tail + ROADMAP M7's <500ms latency.
/// On `RecvError::Lagged` the bridge emits this gap signal; the webview's
/// activity store reconciles by re-querying the durable `activity_log` (page 0,
/// the source of truth) and dedup-merging, so no durable row is lost. The
/// 500ms-typical path stays event-driven via `activity:new`; this fires only on
/// the rare lag.
pub const EVENT_ACTIVITY_LAGGED: &str = "activity:lagged";
/// `account:needs_reauth` - a refresh token was revoked (payload:
/// `{ account_id, email }`, SPEC s11.7).
pub const EVENT_ACCOUNT_NEEDS_REAUTH: &str = "account:needs_reauth";
/// `oauth:complete` - an in-flight add-account / reauth OAuth flow reached a
/// terminal state (payload: `{ session_id, status }`, SPEC s11.7).
///
/// M6: the wizard subscribes so it can advance past the OAuth step without
/// polling. Emitted by the accounts command layer once
/// [`driven_drive::google::oauth::run_pkce_loopback_flow`] resolves; the
/// emit helper lands with that implementer, so the constant is defined-but-
/// uncalled in the M6 scaffold.
#[allow(dead_code)]
pub const EVENT_OAUTH_COMPLETE: &str = "oauth:complete";
/// `updater:available` - a newer release is available (payload: `UpdateInfo`,
/// SPEC s11.7).
///
/// M9a: the About tab + in-app banner subscribe. The periodic updater check
/// (startup + every 6h) and the manual `check_for_update` IPC emit it via
/// [`emit_updater_available`] when [`crate::updater`] finds a newer signed
/// release on the active channel.
pub const EVENT_UPDATER_AVAILABLE: &str = "updater:available";
/// `updater:download_progress` - a staged update is downloading (payload:
/// `{ downloaded, total }`, SPEC s15.2). M9a: emitted by `install_update` on
/// every chunk so the in-app banner shows a progress bar.
pub const EVENT_UPDATER_DOWNLOAD_PROGRESS: &str = "updater:download_progress";
/// `updater:downloaded` - the available update finished downloading and is
/// ready to install (payload: `UpdateInfo`, SPEC s11.7). M9a: emitted by
/// `install_update` right before the relaunch.
pub const EVENT_UPDATER_DOWNLOADED: &str = "updater:downloaded";
/// `restore:progress` - a restore job advanced (payload: `RestoreJobStatus`,
/// SPEC s11.7).
///
/// M8 (restore browser): the background restore job emits this on every per-file
/// progress tick + every file/job state transition so the Restore view renders
/// live progress (per-file + overall, with errors + a terminal `done` state). The
/// same snapshot is recorded on `AppState` so `get_restore_job` can serve a late
/// subscriber.
pub const EVENT_RESTORE_PROGRESS: &str = "restore:progress";

/// Broadcast `sync:status_changed` with the global status payload (SPEC s11.7).
///
/// Thin wrapper over the v2 [`Emitter::emit`] so the orchestrator-event bridge
/// cannot typo the channel name. The payload is whatever the IPC layer hands
/// in (`GlobalSyncStatus` in production); kept generic so the bridge can emit
/// before the concrete DTO crate exists.
pub fn emit_sync_status_changed<P: Serialize + Clone>(
    app: &AppHandle,
    status: &P,
) -> tauri::Result<()> {
    app.emit(EVENT_SYNC_STATUS_CHANGED, status)
}

/// Broadcast `activity:new` with the new activity entry (SPEC s11.7).
///
/// M7 (activity dashboard): the event bridge calls this on every
/// `OrchestratorEvent::ActivityWritten`, forwarding the carried `ActivityEntry`
/// (already the camelCase wire shape) to the webview's live tail.
pub fn emit_activity_new<P: Serialize + Clone>(app: &AppHandle, entry: &P) -> tauri::Result<()> {
    app.emit(EVENT_ACTIVITY_NEW, entry)
}

/// Broadcast `activity:lagged` - the live-tail dropped events on broadcast lag
/// (M7-P1-1, SPEC s11.7). Carries no payload: it is purely a gap signal telling
/// the webview store to reconcile from the durable `activity_log`. `skipped` is
/// the broadcast's reported drop count, attached as a structured field for
/// diagnostics only (the store does not need it to reconcile).
pub fn emit_activity_lagged(app: &AppHandle, skipped: u64) -> tauri::Result<()> {
    app.emit(
        EVENT_ACTIVITY_LAGGED,
        serde_json::json!({ "skipped": skipped }),
    )
}

/// Broadcast `account:needs_reauth` for `account_id` / `email` (SPEC s11.7).
///
/// Emits the `{ account_id, email }` payload the webview's re-consent banner
/// subscribes to (SPEC s11.7 table). Built as an inline JSON object so the
/// shape matches the spec without a dedicated DTO type.
pub fn emit_account_needs_reauth(
    app: &AppHandle,
    account_id: &str,
    email: &str,
) -> tauri::Result<()> {
    app.emit(
        EVENT_ACCOUNT_NEEDS_REAUTH,
        serde_json::json!({ "account_id": account_id, "email": email }),
    )
}

/// Broadcast `updater:available` with the available-update info (SPEC s11.7,
/// s15.2). M9a: the periodic check + the manual `check_for_update` IPC call this
/// when [`crate::updater`] finds a newer signed release on the active channel.
pub fn emit_updater_available<P: Serialize + Clone>(
    app: &AppHandle,
    info: &P,
) -> tauri::Result<()> {
    app.emit(EVENT_UPDATER_AVAILABLE, info)
}

/// Broadcast `updater:download_progress` (SPEC s15.2): `downloaded` of `total`
/// bytes staged so far. `total` is `None` until the server reports a content
/// length. M9a: `install_update` calls this on every chunk so the banner can
/// render a progress bar.
pub fn emit_updater_download_progress(
    app: &AppHandle,
    downloaded: u64,
    total: Option<u64>,
) -> tauri::Result<()> {
    app.emit(
        EVENT_UPDATER_DOWNLOAD_PROGRESS,
        serde_json::json!({ "downloaded": downloaded, "total": total }),
    )
}

/// Broadcast `updater:downloaded` (SPEC s11.7): the staged update finished
/// downloading and is about to be applied + the app relaunched. M9a:
/// `install_update` calls this right before `app.restart()`.
pub fn emit_updater_downloaded<P: Serialize + Clone>(
    app: &AppHandle,
    info: &P,
) -> tauri::Result<()> {
    app.emit(EVENT_UPDATER_DOWNLOADED, info)
}

File: ui/src/ipc/events.ts
// Typed `listen` helpers, one per Rust -> webview event (SPEC s11.7). Each
// helper pins the canonical channel name (matching the constants in
// `src-tauri/src/events.rs`) and types the payload, returning the
// `UnlistenFn` the caller invokes on unmount. `@tauri-apps/api/event`'s
// `listen` is the seam vitest mocks.

import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type {
  AccountSyncStatus,
  ActivityEntry,
  ExclusionPreviewBatch,
  ExclusionPreviewDone,
  ExclusionPreviewError,
  ExecProgress,
  GlobalSyncStatus,
  PauseState,
  RestoreJobStatus,
  UpdateInfo,
  IoSample,
} from "./types";

/** The live payload of `sync:status_changed`. SPEC s11.7 documents the aggregate
 * `GlobalSyncStatus`, but the backend CURRENTLY emits a SINGLE-account snapshot
 * per orchestrator transition (src-tauri/src/assembly.rs `AccountSyncStatusEvent`
 * = `{ account_id, state }`, i.e. an `AccountSyncStatus`); the aggregate shape is
 * reserved for a later milestone. Consumers must handle BOTH and discriminate on
 * the presence of the `accounts` array. */
export type SyncStatusChangedPayload = GlobalSyncStatus | AccountSyncStatus;

/** `sync:status_changed` payload: a per-account `AccountSyncStatus` snapshot today
 * (the aggregate `GlobalSyncStatus` is reserved; see `SyncStatusChangedPayload`). */
export function onSyncStatusChanged(
  handler: (status: SyncStatusChangedPayload) => void
): Promise<UnlistenFn> {
  return listen<SyncStatusChangedPayload>("sync:status_changed", (e) => handler(e.payload));
}

/** `sync:pause_changed` payload: the active `PauseState`, or null once sync is
 * no longer manually paused. Emitted on pause, on resume, when a timed pause
 * auto-expires, and once at boot when a persisted pause is re-applied - so the
 * paused banner appears and disappears with no refresh. */
export function onPauseChanged(handler: (pause: PauseState | null) => void): Promise<UnlistenFn> {
  return listen<PauseState | null>("sync:pause_changed", (e) => handler(e.payload));
}

/** `sync:source_progress` payload: `{ account_id, source_id, progress }` (SPEC
 * s11.7; mirrors src-tauri/src/assembly.rs `SourceProgressEvent`). snake_case on
 * the wire like the rest of the M5 sync DTOs.
 *
 * One throttled execution-progress tick. The orchestrator enters `executing`
 * with a ZEROED `ExecProgress` and never re-emits that state for the source, so
 * these ticks are the ONLY carrier of the moving counters - the progress store
 * folds them over the state's embedded snapshot to keep the top-of-app bar
 * determinate. */
export interface SourceProgressPayload {
  account_id: string;
  source_id: string;
  progress: ExecProgress;
}

export function onSyncSourceProgress(
  handler: (payload: SourceProgressPayload) => void
): Promise<UnlistenFn> {
  return listen<SourceProgressPayload>("sync:source_progress", (e) => handler(e.payload));
}

/** `sync:io_throughput` payload: one live throughput sample (camelCase, the
 * same `IoSample` shape `io_throughput_series` returns). Emitted ~1/s while
 * any disk/network backup work is moving; suppressed (after one trailing
 * zero) while fully idle. */
export function onSyncIoThroughput(handler: (sample: IoSample) => void): Promise<UnlistenFn> {
  return listen<IoSample>("sync:io_throughput", (e) => handler(e.payload));
}

/** `activity:new` payload: ActivityEntry (SPEC s11.7). The Activity dashboard's
 * live tail subscribes to this and prepends new entries (deduped by id). */
export function onActivityNew(handler: (entry: ActivityEntry) => void): Promise<UnlistenFn> {
  return listen<ActivityEntry>("activity:new", (e) => handler(e.payload));
}

/** `activity:lagged` payload (mirrors src-tauri `emit_activity_lagged`): the
 * number of `activity:new` events the bounded broadcast dropped. */
export interface ActivityLaggedPayload {
  skipped: number;
}

/** `activity:lagged` - the live-tail broadcast lagged and dropped one or more
 * `activity:new` events (M7-P1-1 / R1-P1-2, SPEC s11.7). The payload carries the
 * dropped `skipped` count so the store can reconcile ENOUGH pages to cover the
 * gap (not just page 0): the activity store reconciles by re-querying the
 * durable `activity_log` and dedup-merging, so no durable row is lost. */
export function onActivityLagged(
  handler: (payload: ActivityLaggedPayload) => void
): Promise<UnlistenFn> {
  return listen<ActivityLaggedPayload>("activity:lagged", (e) => handler(e.payload));
}

/** `account:needs_reauth` payload: { account_id, email } (SPEC s11.7). */
export interface NeedsReauthPayload {
  account_id: string;
  email: string;
}

export function onAccountNeedsReauth(
  handler: (payload: NeedsReauthPayload) => void
): Promise<UnlistenFn> {
  return listen<NeedsReauthPayload>("account:needs_reauth", (e) => handler(e.payload));
}

/** `oauth:complete` payload: { session_id, status } (SPEC s11.7). */
export interface OAuthCompletePayload {
  session_id: string;
  status: unknown;
}

export function onOauthComplete(
  handler: (payload: OAuthCompletePayload) => void
): Promise<UnlistenFn> {
  return listen<OAuthCompletePayload>("oauth:complete", (e) => handler(e.payload));
}

/** `updater:available` payload: UpdateInfo (SPEC s11.7). */
export function onUpdaterAvailable(handler: (info: UpdateInfo) => void): Promise<UnlistenFn> {
  return listen<UpdateInfo>("updater:available", (e) => handler(e.payload));
}

/** `updater:download_progress` payload: { downloaded, total } (SPEC s15.2). M9a:
 * the in-app banner subscribes to render a progress bar while `installUpdate`
 * stages the update. `total` is null until the server reports a content length. */
export interface UpdaterDownloadProgressPayload {
  downloaded: number;
  total: number | null;
}

export function onUpdaterDownloadProgress(
  handler: (payload: UpdaterDownloadProgressPayload) => void
): Promise<UnlistenFn> {
  return listen<UpdaterDownloadProgressPayload>("updater:download_progress", (e) =>
    handler(e.payload)
  );
}

/** `updater:downloaded` payload: UpdateInfo (SPEC s11.7). */
export function onUpdaterDownloaded(handler: (info: UpdateInfo) => void): Promise<UnlistenFn> {
  return listen<UpdateInfo>("updater:downloaded", (e) => handler(e.payload));
}

/** `restore:progress` payload: RestoreJobStatus (SPEC s11.7). The Restore view's
 * store subscribes to this for live per-file + overall progress, errors, and the
 * terminal `done` state. */
export function onRestoreProgress(
  handler: (status: RestoreJobStatus) => void
): Promise<UnlistenFn> {
  return listen<RestoreJobStatus>("restore:progress", (e) => handler(e.payload));
}

/** `exclusion_preview:batch` payload: a slice of the streaming exclusion
 * preview's walk - the nodes discovered since the last batch plus the running
 * totals. The exclusion editor's tree subscribes to this and DISCARDS any batch
 * whose `previewId` is not the generation it is currently showing. */
export function onExclusionPreviewBatch(
  handler: (batch: ExclusionPreviewBatch) => void
): Promise<UnlistenFn> {
  return listen<ExclusionPreviewBatch>("exclusion_preview:batch", (e) => handler(e.payload));
}

/** `exclusion_preview:done` payload: the exact final totals of a streaming
 * exclusion preview, plus whether it was cancelled rather than completed. */
export function onExclusionPreviewDone(
  handler: (done: ExclusionPreviewDone) => void
): Promise<UnlistenFn> {
  return listen<ExclusionPreviewDone>("exclusion_preview:done", (e) => handler(e.payload));
}

/** `exclusion_preview:error` payload: a streaming preview that had already been
 * handed a generation id failed to set itself up (see `ExclusionPreviewError`).
 * Discarded, like a batch, when it is not the generation on screen. */
export function onExclusionPreviewError(
  handler: (error: ExclusionPreviewError) => void
): Promise<UnlistenFn> {
  return listen<ExclusionPreviewError>("exclusion_preview:error", (e) => handler(e.payload));
}
